src/Coordination/KeeperStateMachine.cpp

```diff
@@ -291,14 +295,91 @@ bool KeeperStateMachine::preprocess(const KeeperStorage::RequestForSession & req
     return true;
 }
 
+KeeperStorage::ResponseForSession KeeperStateMachine::processReconfiguration(
+    const KeeperStorage::RequestForSession& request_for_session)
+{
+    ProfileEvents::increment(ProfileEvents::KeeperReconfigRequest);
```

Comment:
Okay, the reconfig is still forwarded to `app_log`.
what I had in mind was to catch it in KeeperDispatcher in `requestThread` and redirect it to `processReconfiguration` there.

With current approach we still have a problem with replays. Currently you check:
```
    if (!dispatcher.clusterUpdateQueueEmpty())
        return bad_request(ZRECONFIGINPROGRESS);
```

But this can produce different results depending on many different factors like network health (first time we apply `add 2`, server 2 joins instantly and the second reconfigure is processed). After restart there is slower network and it takes time for server 2 to join so the second reconfigur is skipped with `ZRECONFIGINPROGRESS`.

Also, when a node becomes a leader, it search in logs for `conf` log with largest `log_idx` and stores it again, this can undo or create many inconsistent states.

I propose 2 options for reconfig result data:
- instantly return when we push the diff to queue and return the current cluster serialized WITHOUT the diff applied. User can do `get` request on that node to see if the update was applied.
- a bit more complex, but there is `NewConfig` callback in NuRaft that can be used in `KeeperServer` in which we push to the response queue the new cluster serialized OR even better override `commit_conf`. This can break some guarantees from ZK (it will not be ordered with other writes) but I don't think we should care that much about it.

I prefer the first option as it's simpler and makes more sense for async approach. In both solutions we lose persistence slightly but it's much safer and the difference in persistence is minimal (instantly appending vs confirming first that the node is available before appending)

Reply:
Ok, makes sense. Will use the first approach